106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,488 @@
+Doing this will provide us with the following advantages:
+* we will ensure that the operator is in control of when rebalances will be triggered.
+* using the existing `KafkaRebalance` CR system make it easier for users to see what is happening and when, which (as we don't support the Cruise Control UI) enhances observability and will also aids in debugging.
+
+### `imbalance` mode in Strimzi's auto-rebalancing feature
+
+The [`auto-rebalancing`](https://strimzi.io/docs/operators/latest/deploying#proc-automating-rebalances-str) feature in Strimzi allows the operator to run a rebalance automatically when a Kafka cluster is scaled up (by adding brokers) or scaled down (by removing brokers).
+
+Auto-rebalancing in Strimzi currently supports two modes:
+* add-brokers - auto-rebalancing on scale up
+* remove-brokers - auto-rebalancing on scale down
+
+To leverage the automated rebalance on imbalanced cluster (those with detected Goal violations), we will be introducing a new mode to the auto-rebalancing feature.
+The new mode will be called `imbalance`, which means that cluster imbalance was detected and rebalancing should be applied to the all the brokers.
+The mode is defined by setting the `spec.cruiseControl.autoRebalance.mode` field as `imbalance` and the corresponding rebalancing configuration is defined as a reference to a "template" `KafkaRebalance` custom resource, by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).
+This field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration (i.e. the same used for unmodified manual `KafkaRebalance` invocations).
+To provide users more flexibility, they only have to configure the auto-rebalance modes they wish to customise.
+They don't require to set up all the modes and can enable the modes they require.
+They can configure auto-rebalance to enable only for their specific case i.e. setting only `imbalance` mode or other scaling related modes.
+Once the auto-rebalance with `imbalance` mode is enabled, the operator will be ready to trigger auto-rebalance whenever the cluster becomes imbalanced.
+To trigger the auto-rebalance, the operator must know that the cluster is imbalanced due to some goal violation anomaly. 
+We will create our own custom notifier named `StrimziCruiseControlNotifier` to do the same.
+This notifier's job will be to update the operator regarding the goal violations so that the operator can trigger a rebalance (see section [AnomalyDetectorNotifier](./106-auto-rebalance-on-imbalanced-clusters.md#anomalydetectornotifier)).
+With this proposal, we are only going to support auto-rebalance on imbalanced cluster.
```

Comment:
I am not sure I understand, does this mean that the operator will only trigger a partition rebalance for goal violations that don't require manual intervention?